src/main/scala/mesosphere/marathon/tasks/TaskTracker.scala

```diff
@@ -50,60 +51,102 @@ class TaskTracker @Inject() (
 
   def take(appId: PathId, n: Int): Set[MarathonTask] = getTasks(appId).iterator.take(n).toSet
 
-  def created(appId: PathId, task: MarathonTask): Unit = {
-    // Keep this here so running() can pick it up
-    // FIXME: Should be persisted here for task reconciliation
-    //        Wont fix for now since this should be completely remodeled in #462
-    getInternal(appId) += (task.getId -> task)
+  /**
+    * Create a new task and store it in-memory and in the underlying storage.
+    *
+    * If the task exists already, the returned Future will fail.
+    */
+  def created(appId: PathId, task: MarathonTask): Future[MarathonTask] = {
+    val taskId = task.getId
+    getInternal(appId).putIfAbsent(task.getId, task) match {
+      case Some(existingTask) =>
+        Future.failed(new IllegalStateException(s"cannot create task [$taskId] of app [$appId], it already exists"))
+      case None =>
+        repo.store(task).map { task =>
+          log.debug(s"created task [$taskId] of app [$appId]")
+          task
+        }
+    }
   }
 
-  def running(appId: PathId, status: TaskStatus): Future[MarathonTask] = {
+  /**
+    * Process a status update for an existing task and either updates the tasks or removes
+    * it from the TaskTracker.
+    *
+    * If the task does not exist yet, the returned Future will fail.
+    */
+  def statusUpdate(appId: PathId, status: TaskStatus): Future[_] = {
     val taskId = status.getTaskId.getValue
-    getInternal(appId).get(taskId) match {
-      case Some(oldTask) if !oldTask.hasStartedAt => // staged
-        val task = oldTask.toBuilder
-          .setStartedAt(System.currentTimeMillis)
-          .setStatus(status)
-          .build
-        getInternal(appId) += (task.getId -> task)
-        store(appId, task).map(_ => task)
-
-      case Some(oldTask) => // running
-        val msg = s"Task for ID $taskId already running, ignoring"
-        log.warn(msg)
-        Future.failed(new Exception(msg))
+    val maybeAppTask: Option[(InternalApp, MarathonTask)] =
+      for { app <- apps.get(appId); task <- app.tasks.get(taskId) } yield (app, task)
```

Comment:
As described in the documentation of this method, taskUpdate will fail if the task does not exist yet. So it should never create an InternalApp or Task. In fact, it can't because it wouldn't have all necessary information.

Since we did neither introduce nor change apps and getInternal in this patch, I would like to do that in another patch (but I'll take care of that).

How should we rename `apps` and `getInternal`? 

* `apps` -> `appTasks` ?
* `getInternal` -> `getOrCreateAppTasks` ?